### PR TITLE
Added docker-support.bsh + tests

### PR DESCRIPTION
**Pull request by rbuckland**
This docker-support operates on a fairly unobtrusive method of 

```
  - ./ci/docker-support.bsh setup
  # the docker-support setup will set some vars for us that we can just import to this shell
  - . ./ci-env-vars.sh
  - docker build -t ${DOCKER_TAG_NAME} .
  - ./ci/docker-support.bsh publish
```

It expects the following environment variables are set

```
    REGISTRY_HOST: has a built-in default if the CI is travis
    REGISTRY_USERNAME: $$REGISTRY_USERNAME
    REGISTRY_PASSWORD: $$REGISTRY_PASSWORD
    DOCKER_IMAGE=user/image-name
    GIT_EMAIL="ci-email"
    GIT_USERNAME="ci-username"
```

For drone usage, we will default (as globals) 

```
    REGISTRY_HOST
    REGISTRY_USERNAME
    REGISTRY_PASSWORD
    GIT_EMAIL
    GIT_USERNAME
```

That way, the only environment var required is

```
    DOCKER_IMAGE=user/image-name
```

### CODE REVIEW

File: docker-support.bsh [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+  log "show-vars" dump_vars
+  log "create_ci_vars" create_ci_vars
+fi
+
+if [[ ${MODE} == "publish" ]]; then
+  log "vars-check" check_vars_full
+  log "docker-publish -- docker push ${DOCKER_TAG_NAME}" only_on_release docker push ${DOCKER_TAG_NAME}
+  log "docker-remove" docker rmi ${DOCKER_TAG_NAME}
+  log "git-config" git_config
+  log "site-publish" $IMPORT_PATH/ci-publish-site.bsh _site
```
Comment (stephanh) [inline docker-support.bsh]: Should this be separate in case this is a more complex build and we want to do some other things to the site as well?

File: docker-support.bsh [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+  cat << EOF
+REGISTRY_HOST=${REGISTRY_HOST}
+VERSION=${VERSION}
+DOCKER_IMAGE=${DOCKER_IMAGE}
+DOCKER_TAG_NAME=${DOCKER_TAG_NAME}
+EOF
+
+}
+
+function create_ci_vars() {
```
Comment (stephanh) [inline docker-support.bsh]: Why is this needed?

File: docker-support.bsh (hunk 1 above)
Comment (rbuckland) [inline docker-support.bsh]: Agree. I did umm/aah about having it in there.

File: docker-support.bsh (hunk 2 above)
Comment (rbuckland) [inline docker-support.bsh]: It provides good "build" output, it is best to see it on a working project to get a better "feel" for it.

**Comment by rbuckland** [comment]
@stephanh - I added ``-s`` as an option if you need it to publish the site also. 

```
docker-support.sh publish -s 
```
by default it will not publish the site.

File: docker-support.bsh [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+  chmod 755 ci-env-vars.sh
+  echo "created ci-env-vars.sh"
+  echo "----------------------------------------------------"
+  cat ci-env-vars.sh
+  echo "----------------------------------------------------"
+}
+
+function git_config() {
+  git config --global user.email "${GIT_EMAIL}"
```
Comment (stephanh) [inline docker-support.bsh]: So this is not done by the publish script? Should it be done by the publish script?

File: docker-support.bsh (hunk 2 above)
Comment (stephanh) [inline docker-support.bsh]: What do you mean by good build output?

File: docker-support.bsh [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+function git_config() {
+  git config --global user.email "${GIT_EMAIL}" 
+  git config --global user.name "${GIT_USERNAME}" 
+}
+
+#
+# Setup some default vars
+#
+
```
Comment (stephanh) [inline docker-support.bsh]: Newline between function and comment.

File: docker-support.bsh [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+function setup_default_vars() {
+
+  if [ "${CI_NAME}" = "travis" ]; then
+    REGISTRY_HOST=${REGISTRY_HOST:-commbank-docker-dockerv2-local.artifactoryonline.com}
+  fi
+  # if the user supplied REGISTRY_HOST, it will be set here (but we check that later)
```
Comment (stephanh) [inline docker-support.bsh]: Should that be above the if statement?

File: docker-support.bsh [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+  fi
+
+  new_version=$(Version_Get $(cat VERSION))
+  if [[ -z $new_version ]]; then
+      exit 1
+  fi
+  echo "Version Mapped: $version => $new_version"
+  echo "$new_version" > VERSION
+  VERSION=$new_version
+}
```
Comment (stephanh) [inline docker-support.bsh]: Can we move this into lib-ci or the setup-version scripts so it is just part of that setting up the version?

File: docker-support.bsh [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+  echo "Invalid mode ${MODE}" 1>&2
+  usage
+  exit 1 
+fi
+
+get_cmd_opts "$@"
+
+if [[ ${MODE} == "setup" ]]; then
+  log "vars-check" check_vars_base
+  log "docker-login -- ${REGISTRY_USERNAME}@${REGISTRY_HOST}" docker login -u ${REGISTRY_USERNAME} -p ${REGISTRY_PASSWORD} ${REGISTRY_HOST}
```
Comment (stephanh) [inline docker-support.bsh]: Now that we can do unauthenticated pulls of images it would be good if we didn't log in if we didn't have to.

We still need to log in on travis and for pushes.

File: docker-support.bsh (hunk 6 above)
Comment (rbuckland) [inline docker-support.bsh]: originally it was calling the setup-version.bsh .. but due to "weird" oddities that failed spectacularly. ([[ ! vs [ !  bash negation) 

I will push it into lib-ci and remove it out of setup-version

File: docker-support.bsh (hunk 3 above)
Comment (rbuckland) [inline docker-support.bsh]: no it doesn't (just checked :-) again )

File: docker-support.bsh (hunk 2 above)
Comment (rbuckland) [inline docker-support.bsh]: e.g. below. I added a substantive comment in the BASH script as well.

```
:: create_ci_vars
    created ci-env-vars.sh
    ----------------------------------------------------
    #!/bin/bash
    export REGISTRY_HOST=someserver.foobar
    export VERSION=1.2.3-20160510042843-9aafe58-docker-support
    export DOCKER_IMAGE=build/my-special-docker
    export DOCKER_TAG_NAME=someserver.foobar/build/my-special-docker:1.2.3-20160510042843-9aafe58-docker-support
    ----------------------------------------------------
```

File: docker-support.bsh (hunk 4 above)
Comment (rbuckland) [inline docker-support.bsh]: resolved in 404e7d5

File: docker-support.bsh (hunk 5 above)
Comment (rbuckland) [inline docker-support.bsh]: resolved in 404e7d5

File: docker-support.bsh (hunk 6 above)
Comment (rbuckland) [inline docker-support.bsh]: resolved in 9aafe58

File: docker-support.bsh (hunk 3 above)
Comment (stephanh) [inline docker-support.bsh]: Should it?

**Comment by stephanh** [comment]
:+1:
